system/jlib/jmutex.hpp

```diff
@@ -586,67 +586,88 @@ class jlib_decl Monitor: public Mutex
-                                                writeWaiting--;
-                                                cs.leave();
-                                                return false;
-                                            }
-                                            cs.leave();
-                                        }
-                                    }
-                                    else
-                                        writeSem.wait();
-                                    //NB: waiting and locks adjusted before the signal occurs.
-                                }
+    bool lockRead(bool timed, unsigned timeout)
+    {  
+        cs.enter(); 
+        if (writeLocks == 0) 
+        {
+            readLocks++;
+            cs.leave();
+        }
+        else
+        {
+            readWaiting++;
+            cs.leave();
+            if (timed)
+            {
+                if (!readSem.wait(timeout)) {
+                    cs.enter(); 
+                    if (!readSem.wait(0)) {
+                        readWaiting--;
+                        cs.leave();
+                        return false;
+                    }
+                    cs.leave();
+                }
+            }
+            else
+                readSem.wait();
+            //NB: waiting and locks adjusted before the signal occurs.
+        }
+        return true;
+    }
+    // NB: cs should be locked on entry to this method
+    bool doLockWrite(bool timed, unsigned timeout)
+    {
+        if ((readLocks == 0) && (writeLocks == 0))
+        {
+            writeLocks++;
+            cs.leave();
+        }
+        else
+        {
+            writeWaiting++;
+            cs.leave();
+            if (timed)
+            {
+                if (!writeSem.wait(timeout))
+                {
+                    cs.enter(); 
+                    if (!writeSem.wait(0))
+                    {
+                        writeWaiting--;
+                        cs.leave();
+                        return false;
+                    }
+                    cs.leave();
+                }
+            }
+            else
+                writeSem.wait();
+            //NB: waiting and locks adjusted before the signal occurs.
+        }
 #ifdef _DEBUG
-                                exclWriteOwner = GetCurrentThreadId();
+        exclWriteOwner = GetCurrentThreadId();
 #endif
-                                return true;
-                            }
+        return true;
+    }
+    bool lockWrite(bool timed, unsigned timeout)
+    {
+        cs.enter();
+        return doLockWrite(timed, timeout);
+    }
+    bool changeToWrite(bool timed, unsigned timeout)
+    {
+        cs.enter();
+        if (writeLocks)
+        {
+            cs.leave();
+            return false; // failed because already write locked
+        }
+        if (readLocks)
+            readLocks--;
+        return doLockWrite(timed, timeout); // returns false if timed out
```

Comment:
Should we try the doLockWrite() and then if ok decrement readLocks ?

Reply:
yes, but in fact that introduces a whole in itself. It should not release/reduce the readLocskc count here 1st.
Unfortunately, that makes the implementation trickier in that there may be something in a new waiting state.
I'll revisit.